### PR TITLE
File Metadata Backend, API, Tests [PLAT-1128]

### PR DESCRIPTION
**Pull request by erinspace**
## Purpose

This is one big PR to bring together every piece of file metadata that's been merged so far into the feature branch. Truly a group effort!

## Changes

- Add file metadata models
	- FileMetadataRecord
	- FileMetadataSchema
- File Metadata API endpoints
	- List 
	- Create
	- Download
		- downloads the full jsonschema-compatible version of the file metadata, in a specified format
- Serializers for converting file metadata into the format expected by the schema
	- These aren't traditional serializers in the DRF sense
	- They contain the methods to combine the metadata on the file and node with the user entered metadata to create a JSON blob that would validate using the official datacite schema
- New jsonschemas
	- `datacite.json`, taken from a pre-compiled jsonschema [found elsewhere on github](https://github.com/inveniosoftware/datacite/blob/v1.0.1/datacite/schemas/datacite-v4.0.json)
	 - `user_entered_datacite.json` - used for validating what goes in the `metadata` field for `FileMetadataRecord` 
- Process changes in other files:
	- create a new `FileMetadataRecord` for every new file created
	- add a `FILE_METADATA_UPDATED` `NodeLog` every time that happens
	- When a file is copied, its `FileMetadataRecord`s are copied as well, and populated with the old file's `FileMetadataRecord`'s metadata field.
- Housekeeping/organization changes
	- Move files from `api/users/schemas/` to `api/base/schemas/`
- Migration to add a datacite `FileMetadataRecord` for all existing `OsfStorageFile`s

## QA Notes

Surprisingly, not a lot of QA most likely for this gigantic feature as there is nothing yet user facing aside from API endpoints. 

A basic check that the feature was working would be to check that the `metadata_records` relationship appears in file detail endpoints for old files, files can still be created, and that new files have `metadata_records` and detail views on the API.

In particular, here's what I spot checked on staging 3:

A pre-existing node's file list has the metadata_records relationship: https://api.staging3.osf.io/v2/nodes/ma4qn/files/osfstorage/5be5baaca963fe001513516f/
That relationship link can be followed: https://api.staging3.osf.io/v2/files/5be5b7f1a963fe001513515e/metadata_records/
the download link works as expected: https://api.staging3.osf.io/v2/files/5be5b7f1a963fe001513515e/metadata_records/5c54ac6eacbb1a7f958ce75f/download/
A metadata record was also created for a new file https://api.staging3.osf.io/v2/nodes/3jnrd/files/osfstorage/
Tested in Canada and Germany!

pre-existing preprint: https://api.staging3.osf.io/v2/files/5c51a92d2db37d0016cd11af/metadata_records/

Newly created preprint:
https://api.staging3.osf.io/v2/files/5c54b86872a7c20017a66718/metadata_records/

## Documentation

TBD - none exists yet, and we may not want to have user-facing docs for this yet until we are using it ourselves!

## Side Effects

- A new thing is now created every time a file is created.

## Ticket
https://openscience.atlassian.net/browse/PLAT-1128

### CODE REVIEW

File: api/files/views.py [hunk 1]
```diff
@@ -162,3 +166,85 @@ def get_serializer_context(self):
...
+        return None
+
+    def get_object(self):
+        return utils.get_object_or_error(
+            self.get_file().records.filter(_id=self.kwargs[self.record_lookup_url_kwarg]).select_related('schema', 'file'),
+            request=self.request,
+        )
+
+    def get(self, request, **kwargs):
+        file_type = self.request.query_params.get('export', 'json')
```
Comment (caseyrollins) [inline api/files/views.py]: @erinspace, FYI: I used `export` instead of `format` here because `format` is used by DRF to determine whether to return the browsable api or JSON.
http://www.django-rest-framework.org/topics/browsable-api/#formats

File: osf/metadata/serializers.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+        if osfstorage_file.guids.exists():
+            doc['alternateIdentifiers'] = [
+                {
+                    'alternateIdentifier': DOMAIN + osfstorage_file.guids.first()._id,
+                    'alternateIdentifierType': 'URL'
+                }
+            ]
+
+        funder_name = record.metadata.get('funding_agency')
```
Comment (caseyrollins) [inline osf/metadata/serializers.py]: We need to update this since we allow for a list of `funders`: https://github.com/CenterForOpenScience/osf.io/pull/8721/files#diff-06bf61b1aa6656fdef57a942f33507dbR39

File: osf/metadata/serializers.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        data = json.loads(cls.serialize_json(record))
+        return schema40.tostring(data)
+
+    @classmethod
+    def validate(cls, record):
+        # This method needs to be overridden because the OSF cannot currently
+        # issue DOIs for a file, which is required for datacite schema validation.
+        # Manually add a placeholder for validation until we handle this better.
+        json_data = json.loads(cls.serialize_json(record))
+        placeholder = DOI_FORMAT.format(prefix=DATACITE_PREFIX, guid='placeholder')
```
[the remark this answers is not in the document]
Comment (caseyrollins) [reply osf/metadata/serializers.py]: > Does this mean we just should't use datacite's schema directly?

Product decided to continue with datacite's schema.

File: osf/metadata/utils.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
+from website import settings
+
+
+SUBJECT_SCHEME = 'bepress Digital Commons Three-Tiered Taxonomy'
+
+RESOURCE_TYPE_MAP = {
```
Comment (caseyrollins) [inline osf/metadata/utils.py]: Since this is specific to datacite, should consider renaming `DATACITE_RESOURCE_TYPE_MAP` or moving to the `DataciteRecordSerializer`

File: osf/metadata/utils.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    'Research Tool': 'Text',
+    'Thesis': 'Text',
+    'Other': 'Text',
+    '(:unas)': 'Other'
+}
+
+
+def datacite_format_contributors(contributors):
+    """
+    contributors_list: list of OSFUsers to format
```
Comment (caseyrollins) [inline osf/metadata/utils.py]: Minor: I'm not sure what the exact docstring syntax is, but I think it's something like: 
```
:param contributors list: OSFUsers
:returns list: Contributors serialized according to the Datacite schema
```

File: osf/migrations/0135_merge_20180928_1919.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# -*- coding: utf-8 -*-
```
Comment (caseyrollins) [inline osf/migrations/0135_merge_20180928_1919.py]: Look into: recreate file metadata migrations to get rid of the two merge migrations

File: osf/models/metadata.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+from osf.models import NodeLog
+from osf.models.base import BaseModel, ObjectIDMixin
+from osf.models.metaschema import FileMetadataSchema
+from osf.utils import permissions as osf_permissions
+from osf.utils.datetime_aware_jsonfield import DateTimeAwareJSONField
+from osf.metadata.serializers import serializer_registry
+from website.util import api_v2_url
+
+
+def validate_user_entered_metadata(value):
```
Comment (caseyrollins) [inline osf/models/metadata.py]: How about just `validate_metadata`?

File: osf_tests/test_file_metadata.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        license_detail = {
+            'id': no_license.license_id,
+            'year': '2018',
+            'copyrightHolders': ['Woop', 'Yeah']
+        }
+        set_license(node, license_detail, Auth(node.creator))
+        osf_file.save()
+        node.save()
+
+        # create a guid
```
Comment (caseyrollins) [inline osf_tests/test_file_metadata.py]: @erinspace, I don't think I follow -- why is this necessary?

File: api/base/schemas/user_entered_datacite.json [hunk 9]
```diff
@@ -0,0 +1,56 @@
+{
```
[the remark this answers is not in the document]
Comment (erinspace) [reply api/base/schemas/user_entered_datacite.json]: That's a good idea. As the name stands I can see it being confused for elements of the actual datacite schema, where it is not actually.

File: api/base/schemas/utils.py [hunk 10]
```diff
@@ -7,8 +7,10 @@
 
 here = os.path.split(os.path.abspath(__file__))[0]
 
-def from_json(fname):
-    with open(os.path.join(here, fname)) as f:
+def from_json(fname, path=None):
+    if not path:
+        path = here
```
[the remark this answers is not in the document]
Comment (erinspace) [reply api/base/schemas/utils.py]: It wouldn't be, I don't think! This was added because of `social-schema.json` I believe, which is still in `api/users/schemas/`.

File: osf/models/metadata.py (hunk 7 above)
Comment (erinspace) [reply osf/models/metadata.py]: True, since everything in metadata is user entered!

File: api/base/schemas/utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (caseyrollins) [reply api/base/schemas/utils.py]: Ahh, gotcha. I meant to move them all to api/base/schemas/... whoops.

File: osf/migrations/0134_add_fm_record_to_osfstorage_files.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+def remove_records_from_files(state, schema):
+    FileMetadatRecord = state.get_model('osf', 'filemetadatarecord')
+    FileMetadatRecord.objects.all().delete()
+
+def add_records_to_files_sql(state, schema):
+    FileMetadataSchema = state.get_model('osf', 'filemetadataschema')
+    datacite_schema_id = FileMetadataSchema.objects.filter(_id='datacite').values_list('id', flat=True)[0]
+    sql = """
+    INSERT INTO osf_filemetadatarecord (created, modified, _id, metadata, id, file_id, schema_id)
+    SELECT  NOW(), NOW(), generate_object_id(), '{}', nextval('osf_filemetadatarecord_id_seq'), OSF_FILE.id, %d
```
Comment (caseyrollins) [inline osf/migrations/0134_add_fm_record_to_osfstorage_files.py]: `generate_object_id()` ❤️

File: osf_tests/test_file_metadata.py (hunk 8 above)
Comment (erinspace) [reply osf_tests/test_file_metadata.py]: ...I have no recollection of why I added that, and nothing breaks if I remove it. Weird!

File: osf/metadata/serializers.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+@register(schema_id='datacite')
+class DataciteMetadataRecordSerializer(MetadataRecordSerializer):
+
+    osf_schema = 'user_entered_datacite.json'
+
+    @classmethod
+    def serialize_json(cls, record):
+        osfstorage_file = record.file
+        target = osfstorage_file.target
+        target.reload()
```
[the remark this answers is not in the document]
Comment (caseyrollins) [reply osf/metadata/serializers.py]: @erinspace, adding `osf_file.target.reload()` in the test (after `node.save()`) resolved this! I'll make a PR to your fork with the change.

File: osf/migrations/0135_merge_20180928_1919.py (hunk 6 above)
Comment (caseyrollins) [reply osf/migrations/0135_merge_20180928_1919.py]: @erinspace, made a PR to address this: https://github.com/erinspace/osf.io/pull/31

File: osf/migrations/0137_add_fm_record_to_osfstorage_files.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+def remove_records_from_files(state, schema):
+    FileMetadatRecord = state.get_model('osf', 'filemetadatarecord')
+    FileMetadatRecord.objects.all().delete()
+
+def add_records_to_files_sql(state, schema):
+    FileMetadataSchema = state.get_model('osf', 'filemetadataschema')
+    datacite_schema_id = FileMetadataSchema.objects.filter(_id='datacite').values_list('id', flat=True)[0]
+    sql = """
+    INSERT INTO osf_filemetadatarecord (created, modified, _id, metadata, id, file_id, schema_id)
+    SELECT  NOW(), NOW(), generate_object_id(), '{}', nextval('osf_filemetadatarecord_id_seq'), OSF_FILE.id, %d
```
Comment (pattisdr) [inline osf/migrations/0137_add_fm_record_to_osfstorage_files.py]: Do we need to provide the value for the FileMetadataRecord.id? ( nextval('osf_filemetadatarecord_id_seq'))

File: api/files/views.py [hunk 14]
```diff
@@ -165,3 +169,85 @@ def get_serializer_context(self):
...
+        record = self.get_object()
+        try:
+            response = FileResponse(ContentFile(record.serialize(format=file_type)))
```
Comment (pattisdr) [inline api/files/views.py]: So I understand, this is where we return the full serialization of the FileMetadataRecord - with all the items pulled off the target/file + user-added metadata items? Only place is a download link?

File: osf/metadata/serializers.py [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+            'titles': [
+                {
+                    'title': osfstorage_file.name
+                },
+                {
+                    'title': target.title,
+                    'titleType': 'AlternativeTitle'
+                }
+            ],
+            'publisher': 'Open Science Framework',
```
Comment (pattisdr) [inline osf/metadata/serializers.py]: What about serializing preprint file metadata for a branded preprint?  Trying to serialize a psyarxiv preprint file, for example.

File: osf/migrations/0137_add_fm_record_to_osfstorage_files.py [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('osf', '0136_add_datacite_file_metadata_schema'),
+    ]
+
+    operations = [
+        migrations.RunPython(add_records_to_files_sql, remove_records_from_files),
```
Comment (pattisdr) [inline osf/migrations/0137_add_fm_record_to_osfstorage_files.py]: I didn't test this on a large data dump, but if it's slow, we might try RunSQL instead.

File: website/files/utils.py [hunk 17]
```diff
@@ -28,6 +28,13 @@ def copy_files(src, target_node, parent=None, name=None):
             cloned.versions.add(new_fileversion)
         else:
             cloned.versions.add(*src.versions.all())
+
+        # copy over file metadata records
+        if cloned.provider == 'osfstorage':
+            for record in cloned.records.all():
+                record.metadata = src.records.get(schema__name=record.schema.name).metadata
+                record.save()
```
Comment (pattisdr) [inline website/files/utils.py]: 👍 works well

File: osf/metadata/schemas/datacite.json [hunk 18]
```diff
@@ -0,0 +1,607 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "id": "datacite-v4.0.json",
+  "title": "DataCite v4.0",
+  "description": "JSON representation of the DataCite v4.0 schema.",
```
Comment (pattisdr) [inline osf/metadata/schemas/datacite.json]: Just curious, where did we get this JSON representation and how did we decide on this schema?

File: osf/migrations/0136_add_datacite_file_metadata_schema.py [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+logger = logging.getLogger(__file__)
+
+
+def add_datacite_schema(state, schema):
+    FileMetadataSchema = state.get_model('osf', 'filemetadataschema')
+    with open('osf/metadata/schemas/datacite.json') as f:
+        jsonschema = json.load(f)
+    _, created = FileMetadataSchema.objects.get_or_create(
+        _id='datacite',
```
[the remark this answers is not in the document]
Comment (erinspace) [reply osf/migrations/0136_add_datacite_file_metadata_schema.py]: Yep I see your point! I know there was some reasoning behind having the id be a human-readable string. If I remember correctly, it was because we needed to use the `_id` to find the correct schema in the registry. [Here is the decorator](https://github.com/CenterForOpenScience/osf.io/pull/8721/files#diff-898ca0bd378d7ba622a0e1b2c78f2abfR9) and then [in use](https://github.com/CenterForOpenScience/osf.io/pull/8721/files#diff-898ca0bd378d7ba622a0e1b2c78f2abfR34) -- it was essentially so that each `RecordSerializer` knows which schema to use.  

[Here is some of the original discussion](https://github.com/CenterForOpenScience/osf.io/pull/8594#discussion_r209979214) Hooray for notes! This is like a treasure hunt/puzzle made by past-Platform for future-Platform.

[the remark this answers is not in the document]
Comment (pattisdr) [reply osf/migrations/0136_add_datacite_file_metadata_schema.py]: ah, I see! this makes sense.  it's good as-is for now then. I'm super impressed with the amount of documentation on this feature in general: lots of info in JIRA, lots of info on the subtickets, hooray!

File: api/files/views.py (hunk 1 above)
Comment (Johnetordoff) [reply api/files/views.py]: We need a check here to make sure people don't try something like `?export=wombat`, would be nice to have an allowed_file_metadata_formats settings var or static member on this view.

File: api/files/views.py [hunk 20]
```diff
@@ -165,3 +169,85 @@ def get_serializer_context(self):
...
+        except ValueError as e:
+            raise InvalidFilterValue(detail=str(e))
+        file_name = 'file_metadata_{}.{}'.format(record.schema._id, file_type)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api/files/views.py]: Also the file_type might not be a usable extension, it works for json, xml, csv etc, but there's edge cases like text vs .txt. You can see how we might get `file_metadata_guid0.text` or event `file_metadata_guid0.wombat` we should probably map file_type to extenstion and content-type and condense this.

File: osf/metadata/serializers.py [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+        raise ValueError('Format "{}" is not supported.'.format(format))
+
+
+@register(schema_id='datacite')
+class DataciteMetadataRecordSerializer(MetadataRecordSerializer):
+
+    osf_schema = 'osf_datacite.json'
+
+    @classmethod
+    def serialize_json(cls, record):
```
Comment (Johnetordoff) [inline osf/metadata/serializers.py]: I know I shouldn't criticize because I wrote this method (I think), but this is long and should be broke up if just for clarity sake. We might want to standardize this somewhat because XML etc will be much harder to serialize.

File: osf/metadata/utils.py [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+    'Funding Submission': 'Text',
+    'Journal Article': 'Text',
+    'Lesson': 'Text',
+    'Poster': 'Text',
+    'Preprint': 'Text',
+    'Presentation': 'Text',
+    'Research Tool': 'Text',
+    'Thesis': 'Text',
+    'Other': 'Text',
+    '(:unas)': 'Other'
```
Comment (Johnetordoff) [inline osf/metadata/utils.py]: Nit: Things like a "Poster" or "Research Tool" or "Presentation" aren't necessarily "Text" if there's any possibility it couldn't be completely formatted as a text, we should play it safe and be go with "Other"

File: api/files/views.py (hunk 1 above)
Comment (erinspace) [reply api/files/views.py]: Yep, @caseyrollins had that [covered here](https://github.com/CenterForOpenScience/osf.io/pull/8721/files#diff-898ca0bd378d7ba622a0e1b2c78f2abfR31) -- see [this test](https://github.com/CenterForOpenScience/osf.io/pull/8721/files#diff-75927fb4e67948f8a9b55f939db47937R64)

File: osf/metadata/utils.py (hunk 22 above)
Comment (erinspace) [reply osf/metadata/utils.py]: This was a product decision, see [the file metadata assessment
doc](https://docs.google.com/spreadsheets/d/1A3q6rbD3cE3dxzAsXQRgiadfvIu8fU3hPyg7-YFNPjY/) in the "OSF Resource Types" tab.

File: api/files/views.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (erinspace) [reply api/files/views.py]: Check is in the `MetadataRecordSerializer`! See other comment.

File: osf/metadata/serializers.py (hunk 15 above)
Comment (erinspace) [reply osf/metadata/serializers.py]: This is a great question! I am getting a strong sense of deja-vu from this exact question when minting preprint DOIs with CrossRef and what to put for the "publisher" and OSF being decided on no matter the branded preprint -- I think the reasoning was similar. I will track down documentation/an official stance on this tho!

File: osf/migrations/0137_add_fm_record_to_osfstorage_files.py (hunk 16 above)
Comment (erinspace) [reply osf/migrations/0137_add_fm_record_to_osfstorage_files.py]: This migration used to be 0134, when that was the case here were the benchmarks [from an earlier comment](https://github.com/CenterForOpenScience/osf.io/pull/8721#discussion_r221661404) -- On about 1.3 million files locally the migration ran for 1 min 50 seconds.

Comment (pattisdr) [reply osf/migrations/0137_add_fm_record_to_osfstorage_files.py]: wow, okay!

File: api/files/views.py (hunk 14 above)
Comment (erinspace) [reply api/files/views.py]: I believe that was the intention, yes! I gotta admit I don't 100% remember the reasoning for this. Adding it to my list of notes of questions to dig up answers to!

File: api/files/views.py [hunk 23]
```diff
@@ -165,3 +169,85 @@ def get_serializer_context(self):
...
+            request=self.request,
+        )
+
+    def get(self, request, **kwargs):
+        file_type = self.request.query_params.get('export', 'json')
+        record = self.get_object()
+        try:
+            response = FileResponse(ContentFile(record.serialize(format=file_type)))
+        except ValueError as e:
+            raise InvalidFilterValue(detail=str(e))
```
[the remark this answers is not in the document]
Comment (erinspace) [reply api/files/views.py]: True! Trying out `rest_framework.exceptions.ValidationError` and:

```
{
    "errors": [
        {
            "detail": "Format \"dinosaur\" is not supported for metadata file export."
        }
    ],
    "meta": {
        "version": "2.11"
    }
}
```

File: api/files/serializers.py [hunk 24]
```diff
@@ -423,6 +429,57 @@ def get_download_link(self, obj):
...
+
+    file = RelationshipField(
+        related_view='files:file-detail',
+        related_view_kwargs={'file_id': '<file._id>'},
+    )
+
+    schema = RelationshipField(
+        related_view='schemas:file-metadata-schema-detail',
+        related_view_kwargs={'schema_id': '<schema._id>'},
+    )
```
[the remark this answers is not in the document]
Comment (erinspace) [reply api/files/serializers.py]: Very good point. Hmm! I think this requires a bit more discussion/thinking, adding it to my list of things to ponder.

**Comment by erinspace** [comment]
Pausing here to discuss some great points in person/in some meetings/questions for product!

- Are we planning on issuing DOI's for files in the future?
- `osf/metadata/serializers.py` is always using `'publisher': 'Open Science Framework'` for metadata generation regardless weather or not its a file on a branded preprint service
- where we return the full serialization of the FileMetadataRecord - with all the items pulled off the target/file + user-added metadata items? Only place is a download link?
- In the `FileMetadataRecordSerializer`,  we validate the user entered metadata against `osf_datacite.json`, when updating. However, the schema relationship returned in the `GET` request is the FileMetadataSchema created from  `osf/metadata/schemas/datacite.json`. How is the user supposed to know how to format a request to update the metadata when the schema returned is not the schema they need to adhere to? Since we're not returning the serialization of the record itself, just the user-entered metadata, I would expect the schema to match the metadata.

File: api/files/serializers.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (erinspace) [reply api/files/serializers.py]: OK! Had a good long think about this one. How about this as a proposed solution?

So the `schema` relationship is really referring to the `schema` that's stored on the FileMetadataRecord, and the `metadata` is to the `metadata` field on the model, which is what's being updated by a user (the schema is never updated by the user at all, only metadata).

I realized the problem really is that `validate_metadata` on the serializer does its thing and only returns vague errors, and that what it's validating against isn't really transparent. The user entered metadata schema is stored on the record's serializer (different from API serializers, [they can be found here](https://github.com/CenterForOpenScience/osf.io/pull/8721/files#diff-898ca0bd378d7ba622a0e1b2c78f2abfR34)  in the `osf_schema` property. So what if updating the `metadata` field just returned the schema it's validating against as part of the error? Something like:

```
{
    "source": {},
    "meta": {},
    "detail": {
        "message": "Your response of dinosaur for the field related_publication_doi was invalid.",
        "metadata_schema": {
            "description": "This is validation for file metadata fields that the OSF allows to be edited by the user. These fields will match the fields presented to the user on a file metadata form on the OSF.",
            "title": "OSF Specific Fields for the Datacite Schema",
            "$id": "/osf_datacite",
            "additionalProperties": "False",
            "$schema": "http://json-schema.org/draft-07/schema#",
            "type": "object",
            "properties": {
                "related_publication_doi": {
                    "pattern": "10\\..+/.+",
                    "type": "string",
                    "description": "The DOI of a related resource."
                },
                "funders": {
                    "items": {
                        "type": "object",
                        "properties": {
                            "funding_agency": {
                                "type": "string",
                                "description": "The name of the funding provider"
                            },
                            "grant_number": {
                                "type": "string",
                                "description": "The code assigned by the funder to a sponsored award (grant)."
                            }
                        }
                    },
                    "type": "array",
                    "description": "Funders responsible for funding the resource."
                },
                "file_description": {
                    "type": "string",
                    "description": "The description of the file"
                },
                "resource_type": {
                    "enum": [
                        "Audio/Video",
                        "Dataset",
                        "Image",
                        "Model",
                        "Software",
                        "Book",
                        "Funding Submission",
                        "Journal Article",
                        "Lesson",
                        "Poster",
                        "Preprint",
                        "Presentation",
                        "Research Tool",
                        "Thesis",
                        "Other"
                    ],
                    "type": "string"
                }
            }
        }
    }
}
```

That way, we're not re-naming fields on a serializer to be different than how they're stored on the model, and the user is still getting feedback on what fields are expected to go into the `metadata` field. 

@pattisdr @sloria what do you think?

[the remark this answers is not in the document]
Comment (pattisdr) [reply api/files/serializers.py]: I think this is a smart idea @erinspace - As you pointed out, osf_datacite.json is the schema for the `metadata` field only, not the schema for the FileMetadataRecord as a whole, which we are serializing here.   I think it makes a lot of sense to return it as part of the error message, so the user can figure out what is wrong with their request, if needed, while still properly serializing the FileMetdataRecord. Win-win!

[the remark this answers is not in the document]
Comment (sloria) [reply api/files/serializers.py]: I'm also good with that. Maybe add `metadata_schema` to `meta` in the error object so that it complies with JSON API: https://jsonapi.org/format/#error-objects

File: osf/migrations/0137_add_fm_record_to_osfstorage_files.py (hunk 13 above)
Comment (erinspace) [reply osf/migrations/0137_add_fm_record_to_osfstorage_files.py]: I think this is already happening!

```
    INSERT INTO osf_filemetadatarecord (created, modified, _id, metadata, id, file_id, schema_id)
    SELECT  NOW(), NOW(), generate_object_id(), '{}', nextval('osf_filemetadatarecord_id_seq'), OSF_FILE.id, %d
```

But maybe I am missing something?

Comment (pattisdr) [reply osf/migrations/0137_add_fm_record_to_osfstorage_files.py]: geez, I wasn't clear at all, I meant, can we take it away? I thought ids were inserted automatically.

File: osf/metadata/serializers.py (hunk 15 above)
Comment (erinspace) [reply osf/metadata/serializers.py]: Updating here for clarity -- discussed in a meeting, leaving it this way so that it matches CrossRef's metadata for preprints.

File: osf/migrations/0137_add_fm_record_to_osfstorage_files.py (hunk 13 above)
Comment (erinspace) [reply osf/migrations/0137_add_fm_record_to_osfstorage_files.py]: Oh oh oh I see YES! Tried removing that and it worked, thank you!

File: api_tests/files/views/test_file_metadata_record_detail.py [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+        assert res.json['data']['id'] == preprint_record._id
+
+        # unauthenticated cannot view unpublished preprint file metadata record
+        res = app.get(self.get_url(unpublished_preprint_record), expect_errors=True)
+        assert res.status_code == 401
+
+        # authenticated contributor can view unpublished preprint file metadata record
+        res = app.get(self.get_url(unpublished_preprint_record), auth=user.auth)
+        assert res.status_code == 200
+        assert res.json['data']['id'] == unpublished_preprint_record._id
```
Comment (pattisdr) [inline api_tests/files/views/test_file_metadata_record_detail.py]: yay preprint record tests 👍

**Comment by pattisdr** [comment]
going to staging3!

File: osf/migrations/0137_add_fm_record_to_osfstorage_files.py (hunk 16 above)
Comment (pattisdr) [reply osf/migrations/0137_add_fm_record_to_osfstorage_files.py]: Matt says the migration took 30 min on prod data which is pushing it. Looking into optimizing this.
